### DISCUSSION TITLE
Add `enqueue` Method to Entrypoint Decorator

### DISCUSSION THREAD

**Discussion by janbjorge** [janbjorge/pgqueuer#220] [Ideas]
Enhance the `entrypoint` decorator to add an `enqueue` method to the decorated functions. This will allow developers to enqueue jobs directly from the entrypoint function, reducing repetitive code and simplifying job management. The new method should incorporate the existing `enqueue` logic from the `queries` class, allowing seamless integration and reuse of the current functionality. The new method should be implemented in a type-safe manner to maintain consistency and reliability. While this feature may not be suitable for all users, it could significantly simplify life for those who need to enqueue jobs within the same process.

### Feature Summary:
- **Decorator Update**: Modify the `entrypoint` decorator to automatically attach an `enqueue` method to decorated functions. This method will enable direct job enqueuing, simplifying job queue management by utilizing the current `enqueue` logic in the `queries` class.
- **Type Safety**: Ensure the `enqueue` method is type-safe using Python's `Protocol` and `cast()` for proper type-checking support. This ensures compatibility with tools like MyPy and provides reliable type hints in IDEs.
- **Ease of Use**: Adding the `enqueue` method will allow developers to enqueue jobs with minimal code, making job processing functions more self-contained.

### Usage Example:
With the updated decorator, the usage would look like this:
  ```python
  @pgq.entrypoint("fetch")
  async def process_message(job: models.Job) -> None:
      print(f"Processed message: {job!r}")

  # Enqueue a new job directly using the decorated function
  await process_message.enqueue(
      payload=b"example_payload",
      priority=1,
      execute_after=timedelta(seconds=5)
  )  # Example of directly enqueuing a job
  ```
  This update makes the decorated function more versatile, enabling both execution and enqueuing in a unified and type-safe manner.

### Acceptance Criteria:
- [ ] `enqueue` method is attached to decorated functions, allowing direct enqueuing with equivalent capabilities to the existing `queries.enqueue()` method.
- [ ] Correct type annotations for both the `__call__` and `enqueue` methods to ensure proper function signature representation.
- [ ] Type checking with tools like MyPy should pass without issues, verifying that both the original entrypoint and the added `enqueue` method are fully type-safe.
- [ ] Example usage is documented and verified to work as expected in different scenarios, including handling multiple jobs at once.

**Comment by daaain** [comment]
It does look nicer than hardcoded strings, but I've already been using constants in the decorator to name the entrypoint and reference later, so wouldn't make that much difference 🤷 

Wait, but also, why do you need to pass in `entrypoint`, would the function reference be enough?

**Reply by janbjorge** [reply]
I use enums instead of strings, which makes the code a bit less error-prone. I also have a separate `Enqueue` class that includes queue functions for each entry point. This setup has made my life at least a bit easier.

The other day, I used [[lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache)](https://docs.python.org/3/library/functools.html#functools.lru_cache), and I noticed that it adds a `.clear_cache()` method to the functions it decorates. It made me wonder if adding something similar to PGQueuer would be useful. I'm not entirely sure if it makes sense here, though. It looks nice, but I'm unsure if it actually adds any value or convenience—especially since, in most cases, enqueue and dequeue operations are "far apart."

> Wait, but also, why do you need to pass in `entrypoint`? Would the function reference be enough?

That was a major oversight—thanks for pointing it out. (I've updated it.)
